### PR TITLE
fix: mb_convert_encoding PHP 8.2 deprecation

### PR DESCRIPTION
**Pull request by shvlv**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Fix.

**What is the current behavior?** (You can also link to an open issue here)
I added minor improvements on the top of @nlemoine's PR https://github.com/inpsyde/assets/pull/49. It was inspired by very good SO answer - https://stackoverflow.com/a/8218649/13962131 and we tested same change for several projects already where we need HTML processing.

So the problem is (https://www.php.net/manual/en/function.mb-convert-encoding.php):
> from 8.2.0 mb_convert_encoding() will no longer return the following non text encodings: "Base64", "QPrint", "UUencode", "HTML entities", "7 bit" and "8 bit".

**What is the new behavior (if this is a feature change)?**

The right way to do things is `mb_encode_numericentity`.

`DOMDocument::loadHTML` loads ISO-8859-1 by default. So for `map` argument, we choose:
* start_code - `128` (`0x80`). Anything above ISO-8859-1 that supports `0-127`.
* end_code - I think using the maximum Unicode value `U+10FFFF` is better. 
* offset - `0` as we don't need to change code
* mask -  we don't need to change anything so that we can repeat the end_code value. But I like the idea of `~0` that produces the maximal byte for your system. i.e. `1111111111111111111111111111111111111111111111111111111111111111` for a 64bit system. It guarantees the mask never changes its actual value.

Regarding `html_entity_decode`, I think it's an important addition because we can leave HTML entities for HTML content, and the browser handles them as expected. But we don't want to break Unicode characters inside JS. So if the user has `var title = 'Lösungen'`, changing to `var = 'L&#246;sungen'` is not good.

Also, tests are added, including the character from the Unicode top plate - https://util.unicode.org/UnicodeJsps/character.jsp?a=2000B. I.e. `&#131083;`

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

**Other information**:

### CODE REVIEW

**Comment by nlemoine** [comment]
Thanks you @shvlv, this solves PHP 8.2 deprecations.

However, I stumbled upon another bug today. Here's the context.

I'm using Vite to build my assets (with a custom Vite loader). In dev mode, I have to include vite client to handle hot reload, etc.

The default URL for vite server is  `http://[::1]:5173/path/to/build/@vite/client`. When this goes through DomDocument, the URL ends up encoded like so:
```html
"http://%5B::1%5D:5173/path/to/build/@vite/client
```

`html_entity_decode` doesn't take effect here. I couldn't find a solution to this issue.

I was wondering if instead of DomDocument (which has a lot of quirks and can be heavy on memory) and since the only manipulations achieved here targets attributes, you could leverage [the new Tag processor API](https://make.wordpress.org/core/2023/03/07/introducing-the-html-api-in-wordpress-6-2/) introduced in 6.2?

Of course, that would mean a WP 6.2 requirement on the other hand. 

6.3 also [introduced native support](https://make.wordpress.org/core/2023/07/14/registering-scripts-with-async-and-defer-attributes-in-wordpress-6-3/) for `defer` and `async`.

**Comment by shvlv** [comment]
After internal discussion, the new approach with WordPress HTML API was implemented. The following points should be discussed additionally:
1. Do we consider the current change as a breaking change, or will triggering deprecation be enough?
2. We need integration tests for this change. I added database-less tests as PoC. Maybe we can consider a different approach here.

**Comment by nlemoine** [comment]
Quick note: from 6.5, there is now a [`wp_enqueue_script_module`](https://github.com/WordPress/wordpress-develop/blob/18db904c293f43f5e928b74835d0969cd60c1e2e/src/wp-includes/script-modules.php#L68-L103) function.

**Comment by nlemoine** [comment]
@shvlv Friendly bump, any progress on this?

**Comment by shvlv** [comment]
@nlemoine Sorry for the delay, we have a vacation time :smile: will proceed as soon as possible

**Comment by nlemoine** [comment]
Hello @shvlv, would you mind having a look a this PR? 🙂

File: src/OutputFilter/AttributesOutputFilter.php [hunk 1]
```diff
@@ -15,105 +15,55 @@
...
-     *
-     * @return void
-     */
-    protected function applyAttributes(\DOMElement $script, array $attributes)
+    protected function applyAttributes(\WP_HTML_Tag_Processor $script, array $attributes): void
     {
         foreach ($attributes as $key => $value) {
-            $key = esc_attr((string) $key);
-            if ($script->hasAttribute($key)) {
+            $key = esc_attr((string)$key);
```
Comment (tyrann0us) [inline src/OutputFilter/AttributesOutputFilter.php:55]:
```suggestion
            $key = esc_attr((string) $key);
```

File: src/OutputFilter/AttributesOutputFilter.php [hunk 2]
```diff
@@ -15,105 +15,55 @@
...
+            if ((string) $script->get_attribute($key)) {
                 continue;
             }
             if (is_bool($value) && !$value) {
                 continue;
             }
             $value = is_bool($value)
                 ? esc_attr($key)
-                : esc_attr((string) $value);
+                : esc_attr((string)$value);
```
Comment (tyrann0us) [inline src/OutputFilter/AttributesOutputFilter.php:64]:
```suggestion
                : esc_attr((string) $value);
```

**Review by tyrann0us — APPROVED** [review]
Thanks for working on this! I have two comments but they don't stop me from approving.
1. I'm not sure how much we should care about WordPress < 6.2. It might be enough simply returning `$html` without triggering an error message (and have a test for it).
2. If the QA issues in the unchanged files are quick wins, it would be great to have them fixed according to the Boy Scout rules.

**Comment by shvlv** [comment]
> * I'm not sure how much we should care about WordPress < 6.2. It might be enough simply returning `$html` without triggering an error message (and have a test for it).

Done partially. We decided to postpone integration testing for now.

> * If the QA issues in the unchanged files are quick wins, it would be great to have them fixed according to the Boy Scout rules.

It is better to handle it in the standalone PR.
